### PR TITLE
ARROW-6336: [Python] Add notes to pyarrow.serialize/deserialize to clarify that these functions do not read or write the standard IPC protocol

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

File: python/pyarrow/serialization.pxi [hunk 1]
```diff
@@ -334,7 +334,14 @@ cdef class SerializedPyObject:
 
 
 def serialize(object value, SerializationContext context=None):
-    """EXPERIMENTAL: Serialize a Python sequence
+    """EXPERIMENTAL: Serialize a general Python sequence for transient storage
+    and transport. This may have better performance and memory efficiency than
+    Python pickle.
+
+    Note: this function produces data that is incompatible the standard Arrow
```
Comment (jorisvandenbossche) [inline python/pyarrow/serialization.pxi]:
```suggestion
    Note: this function produces data that is incompatible with the standard Arrow
```

File: python/pyarrow/serialization.pxi [hunk 2]
```diff
@@ -334,7 +334,14 @@ cdef class SerializedPyObject:
...
+    IPC binary protocol, .e. it cannot be used with ipc.open_stream or
```
Comment (jorisvandenbossche) [inline python/pyarrow/serialization.pxi]:
```suggestion
    IPC binary protocol, i.e. it cannot be used with ipc.open_stream or
```

File: python/pyarrow/serialization.pxi [hunk 3]
```diff
@@ -403,7 +411,8 @@ def read_serialized(source, base=None):
 
 
 def deserialize_from(source, object base, SerializationContext context=None):
-    """EXPERIMENTAL: Deserialize a Python sequence from a file.
+    """EXPERIMENTAL: Deserialize a Python sequence from a file. This only can
+    interact with data produced by pyarrow.serialize ot pyarrow.serialize_to
```
Comment (jorisvandenbossche) [inline python/pyarrow/serialization.pxi]:
```suggestion
    interact with data produced by pyarrow.serialize or pyarrow.serialize_to
```

File: python/pyarrow/serialization.pxi [hunk 4]
```diff
@@ -443,9 +452,9 @@ def deserialize_components(components, SerializationContext context=None):
 
 
 def deserialize(obj, SerializationContext context=None):
-    """
-    EXPERIMENTAL: Deserialize Python object from Buffer or other Python object
-    supporting the buffer protocol
+    """EXPERIMENTAL: Deserialize Python object from Buffer or other Python
+    object supporting the buffer protocol. This only can interact with data
+    produced by pyarrow.serialize ot pyarrow.serialize_to
```
Comment (jorisvandenbossche) [inline python/pyarrow/serialization.pxi]:
```suggestion
    produced by pyarrow.serialize or pyarrow.serialize_to
```

**Review by jorisvandenbossche — COMMENTED** [review]
I would add a similar note in the "Arbitrary Object Serialization" section of the user guide: https://github.com/apache/arrow/blob/master/docs/source/python/ipc.rst#arbitrary-object-serialization

**Comment by wesm** [comment]
Thanks, will do.

**Comment by wesm** [comment]
Travis CI: https://travis-ci.org/wesm/arrow/builds/587238743

**Pull request closed without merging**